## Experiment 6

### Code Refactoring Exercise: Identify and refactor code snippets to improve code quality, readability, and performance, focusing on techniques such as extracting methods, removing duplication, and improving naming conventions. 

### 1.  Naming conventions, readability and Code Quality

### **Original Code**

Below is an example of a Python function that processes user data. The code has several issues:

- **Poor Naming Conventions**: Variable and function names are unclear.
- **Code Duplication**: Similar validation checks are repeated.
- **Large Function**: The function handles multiple responsibilities, making it hard to read and maintain.
- **Complex Conditional Statements**: Nested `if-else` blocks reduce readability.

**Explanation of the Original Code:**

- **Function Purpose**: The `proc` function processes a user's data by determining their status, calculating discounts based on certain criteria, and computing the total amount.
- **Issues Identified**:
  - **Unclear Variable Names**: Variables like `u`, `a`, `b`, `c`, and `d` do not convey meaningful information.
  - **Duplicated Logic**: The user status checks for 'admin' and 'editor' roles are repetitive.
  - **Single Large Function**: The function handles multiple tasks, violating the Single Responsibility Principle.
  - **Magic Numbers**: The discount rate `0.1` is hard-coded without explanation.

In [3]:
# Original Code: User Data Processing

def proc(u):
    if u['a'] == 'admin' and u['b']:
        print("Admin user active")
    elif u['a'] == 'editor' and u['b']:
        print("Editor user active")
    else:
        print("User inactive")
    
    if u['c'] > 1000:
        discount = u['c'] * 0.1
        print(f"Discount applied: {discount}")
    else:
        print("No discount")
    
    total = u['c'] + u['d'] - discount
    print(f"Total amount: {total}")
    
# Usage
# Sample User Data
user_data = {
    'role': 'admin',
    'is_active': True,
    'income': 1500,
    'bonus': 200
}

# Processing the User
process_user(user_data)

Admin user active
Discount applied: 150.0
Total amount: 1550.0


### **Refactored Code**

The refactored version addresses the issues by:

- **Improving Naming Conventions**: Using descriptive names for variables and functions.
- **Extracting Methods**: Breaking down the function into smaller, reusable functions.
- **Removing Duplication**: Consolidating repetitive logic.
- **Simplifying Conditionals**: Making conditional statements more readable.
- **Using Constants**: Replacing magic numbers with named constants for clarity.

**Explanation of the Refactored Code:**

1. **Constants Defined**:
   - `DISCOUNT_RATE`: Represents the discount percentage, replacing the magic number `0.1`.
   - `HIGH_INCOME_THRESHOLD`: Defines the income threshold for applying discounts, replacing the magic number `1000`.

2. **Main Function (`process_user`)**:
   - **Single Responsibility**: Now only orchestrates the processing by calling other functions.
   - **Early Exit**: If the user is inactive, the function exits early, enhancing readability.

3. **Helper Function (`is_user_active`)**:
   - **Descriptive Naming**: Clearly indicates its purpose.
   - **Consolidated Logic**: Checks if the user's role is in the list of active roles and if the user is active.
   - **Reusability**: Can be reused for other user status checks if needed.

4. **Helper Function (`apply_discount_if_applicable`)**:
   - **Encapsulated Logic**: Handles discount calculation based on income.
   - **No Duplication**: The discount logic is centralized, eliminating repetition.
   - **State Modification**: Updates the `user` dictionary with the calculated discount.

5. **Helper Function (`calculate_total_amount`)**:
   - **Clear Purpose**: Calculates and prints the total amount.
   - **Separation of Concerns**: Keeps calculation logic separate from validation and discount application.

6. **Improved Readability and Maintainability**:
   - The refactored code is easier to read, understand, and maintain.
   - Each function has a clear, singular purpose, adhering to best practices.
   

#### **Benefits of Refactoring Applied**

1. **Enhanced Readability**: Clear and descriptive names make the code self-explanatory.
2. **Improved Maintainability**: Smaller functions are easier to manage and test.
3. **Reduced Duplication**: Centralizing repeated logic minimizes errors and simplifies updates.
4. **Better Performance**: While not a primary focus here, cleaner code can lead to performance optimizations.
5. **Scalability**: The modular structure allows for easier addition of new features or modifications.

In [4]:
# Refactored Code: User Data Processing

DISCOUNT_RATE = 0.1
HIGH_INCOME_THRESHOLD = 1000

def process_user(user):
    if not is_user_active(user):
        print("User inactive")
        return
    
    apply_discount_if_applicable(user)
    calculate_total_amount(user)

def is_user_active(user):
    active_roles = ['admin', 'editor']
    if user['role'] in active_roles and user['is_active']:
        print(f"{user['role'].capitalize()} user active")
        return True
    return False

def apply_discount_if_applicable(user):
    if user['income'] > HIGH_INCOME_THRESHOLD:
        discount = user['income'] * DISCOUNT_RATE
        user['discount'] = discount
        print(f"Discount applied: {discount}")
    else:
        user['discount'] = 0
        print("No discount")

def calculate_total_amount(user):
    total = user['income'] + user['bonus'] - user['discount']
    print(f"Total amount: {total}")

# Usage
# Sample User Data
user_data = {
    'role': 'admin',
    'is_active': True,
    'income': 1500,
    'bonus': 200
}

# Processing the User
process_user(user_data)

Admin user active
Discount applied: 150.0
Total amount: 1550.0


### 2. Reducing Complexity

### Original Code Explanation

The original code snippet is designed to find all prime numbers up to a specified limit (in this case, 10,000). It includes a function to determine if a number is prime and another function to collect all prime numbers within the given range. Here’s a breakdown of its functionality:

#### **Function Purpose**:
The main purpose of this code is to identify and count the total number of prime numbers within a specified limit, while also measuring the execution time of the process.

### **Functionality Breakdown**:

1. **`is_prime(n)`**:
   - This function checks whether a given number `n` is a prime number. It returns `True` if `n` is prime and `False` otherwise.
   - **Logic**:
     - It first checks if `n` is less than or equal to 1, in which case it cannot be prime.
     - It then iterates from `2` to `n-1`, checking if `n` is divisible by any of these numbers. If it finds a divisor, it concludes that `n` is not prime.

   **Example**:
   ```python
   def is_prime(n):
       if n <= 1:
           return False
       for i in range(2, n):
           if n % i == 0:
               return False
       return True
   ```

2. **`find_primes_original(limit)`**:
   - This function generates a list of all prime numbers up to the specified `limit`.
   - **Logic**:
     - It initializes an empty list called `primes`.
     - It then iterates over every number from `0` to `limit - 1`, checking if each number is prime using the `is_prime` function. If a number is found to be prime, it is appended to the `primes` list.

   **Example**:
   ```python
   def find_primes_original(limit):
       primes = []
       for number in range(limit):
           if is_prime(number):
               primes.append(number)
       return primes
   ```

3. **Execution Timing**:
   - The code measures the execution time of the `find_primes_original` function by capturing the start time before execution and the end time afterward. The difference gives the total time taken to find the prime numbers.

   **Example**:
   ```python
   start_time = time.time()
   primes_original = find_primes_original(10000)
   end_time = time.time()
   ```

4. **Output**:
   - Finally, it prints the total number of prime numbers found and the execution time taken to perform the computation.

   **Example**:
   ```python
   print(f"Original Total Prime Numbers: {len(primes_original)}")
   print(f"Execution Time (Original): {end_time - start_time:.6f} seconds")
   ```

### **Issues Identified**:

1. **Inefficient Prime Checking**:
   - The `is_prime` function uses a naive approach to check for primality, iterating through all numbers from `2` to `n-1`. This results in a time complexity of O(n) for each prime check. When combined with the loop in `find_primes_original`, it leads to an overall time complexity of O(n^2) for finding all primes up to `limit`.

   **Recommendation**: Implement a more efficient algorithm, such as the Sieve of Eratosthenes, to reduce the time complexity to O(n log log n).

2. **Redundant Range in Prime Check**:
   - The check in `is_prime` can be optimized to iterate only up to the square root of `n` instead of `n-1`, as any non-prime number must have at least one factor less than or equal to its square root.

   **Recommendation**: Change the loop condition in `is_prime` to iterate from `2` to `int(n**0.5) + 1`.

3. **Poor Naming Conventions**:
   - The function names and variable names could be more descriptive. For instance, `find_primes_original` could be renamed to `find_primes_up_to_limit` to clearly indicate its purpose.

   **Recommendation**: Improve naming conventions to enhance code readability and understanding.

4. **Magic Number**:
   - The number `10000` is hard-coded in the execution block without context, which makes the code less flexible.

   **Recommendation**: Define a constant or allow user input for the limit, improving the code's adaptability for different use cases.
   
### Explanation:

### 1. **Code Logic Efficiency**
   - **Inefficient Prime Checking**:
     - The `is_prime(n)` function employs a naive algorithm to check if a number is prime. It iterates through all integers from `2` to `n-1` to find divisors. This results in a time complexity of O(n) for each individual prime check.
     - Since `find_primes_original(limit)` calls `is_prime(n)` for each number up to `limit`, the combined time complexity becomes O(n^2). This is particularly inefficient for larger values of `limit`, such as 10,000.

   - **Recommendation**: 
     - Implement a more efficient algorithm, such as the **Sieve of Eratosthenes**. This method can find all prime numbers up to a specified limit with a time complexity of O(n log log n), significantly improving performance.

### 2. **Code Logic Optimization**
   - **Redundant Range in Prime Check**:
     - The current implementation of the `is_prime(n)` function checks divisibility from `2` to `n-1`. However, if a number is not prime, it must have at least one factor less than or equal to its square root. 
     - By iterating only up to the square root of `n`, the number of checks can be reduced, leading to better performance.

   - **Recommendation**: 
     - Modify the loop condition in `is_prime` to iterate from `2` to `int(n**0.5) + 1`, thus optimizing the primality check.

### 3. **Code Quality and Readability**
   - **Naming Conventions**:
     - The function and variable names in the code could be more descriptive. For example, `find_primes_original` does not clearly communicate its purpose. Names should indicate functionality and intent for better understanding and maintainability.
     - Poor naming conventions can lead to confusion, especially for new developers or when revisiting the code after some time.

   - **Recommendation**: 
     - Rename `find_primes_original` to something more descriptive, like `find_primes_up_to_limit`, to clearly convey its function. Similarly, consider renaming variables for improved clarity.

### 4. **Magic Numbers**
   - **Hard-Coded Values**:
     - The value `10,000` is hard-coded in the execution block without any context or explanation. This practice, known as using a "magic number," reduces code flexibility and adaptability.
     - When the limit is hard-coded, it limits the function's usability and may lead to errors if the value needs to be changed in multiple places.

   - **Recommendation**: 
     - Define a constant or allow user input for the limit. For example, you could define `LIMIT = 10000` at the beginning of the script. This not only enhances readability but also allows for easy adjustments in the future.

### 5. **Execution Timing Clarity**
   - **Lack of Context for Execution Timing**:
     - While the code measures execution time, it does not provide context or comments explaining why timing is being measured. This can be confusing for someone who is reviewing the code.
     - Additionally, the timing information could be more informative if presented in a more user-friendly format or combined with a message indicating the purpose of the calculation.

   - **Recommendation**: 
     - Add comments explaining the importance of measuring execution time and consider formatting the output to make it clearer. For instance, printing a message that summarizes the total number of prime numbers found alongside the execution time can provide a clearer context for the user.

### 6. **Documentation and Comments**
   - **Insufficient Documentation**:
     - The code lacks adequate comments explaining the purpose of functions and key logic. Good documentation helps future maintainers understand the thought process behind the implementation, making the codebase more accessible.
  
   - **Recommendation**: 
     - Add docstrings to functions and comments within the code to explain the purpose of key sections. For example, a docstring for the `is_prime(n)` function could explain its input, output, and the logic used for determining primality.

### Summary
By addressing these issues, the code can be made more efficient, maintainable, and understandable. Optimizing the algorithm for checking prime numbers, improving naming conventions, eliminating magic numbers, providing clearer execution timing context, and enhancing documentation will collectively lead to a better quality codebase.

In [24]:
import time

def prime_check(n):
    if n <= 1:
        return False
    for j in range(2, n):  # Inefficient loop checking all numbers
        if n % j == 0:
            return False
    return True

def get_primes(lim):
    result = []  # Using a non-descriptive variable name
    for num in range(lim):  # Looping through all numbers
        if prime_check(num):  # Calling prime_check for every number
            result.append(num)  # Appending to result if prime
    return result  # Returning the results

# Timing the execution of the original code
start_t = time.time()  # Poor naming convention
primes_found = get_primes(10000)  # Finding primes up to 10,000
end_t = time.time()  # Poor naming convention

print(f"Total Prime Numbers Found: {len(primes_found)}")  # Non-informative output
print(f"Time Taken: {end_t - start_t:.6f} seconds")  # Non-informative output


Total Prime Numbers Found: 1229
Time Taken: 0.377002 seconds


### Refactored Code Explanation

The refactored code snippet improves the process of finding all prime numbers up to a specified limit (10,000) by optimizing the algorithm used for primality testing. The goal is to enhance both performance and readability. 

#### **Function Purpose**:
The primary purpose of this code is to efficiently identify and count all prime numbers within a specified limit, while measuring the execution time of the process.

### **Functionality Breakdown**:

1. **`is_prime_refactored(n)`**:

#### **Purpose**
The `is_prime_refactored(n)` function determines whether a given number `n` is a prime number. It uses an optimized algorithm to reduce the number of necessary checks, making it significantly faster than the original approach.

### **Function Definition**
```python
def is_prime_refactored(n):
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True
```

### **Step-by-Step Explanation**

Let's dissect each part of the function with examples to illustrate how it works.

#### 1. **Check if `n` is Less Than or Equal to 1**
```python
if n <= 1:
    return False
```
- **Explanation**: By definition, prime numbers are greater than 1. Numbers less than or equal to 1 are not prime.
- **Example**:
  - `n = 1`: `1 <= 1` → **Not prime** (`False`)
  - `n = -5`: `-5 <= 1` → **Not prime** (`False`)

#### 2. **Check if `n` is 2 or 3**
```python
if n <= 3:
    return True
```
- **Explanation**: The numbers 2 and 3 are prime numbers.
- **Example**:
  - `n = 2`: `2 <= 3` → **Prime** (`True`)
  - `n = 3`: `3 <= 3` → **Prime** (`True`)

#### 3. **Eliminate Multiples of 2 and 3**
```python
if n % 2 == 0 or n % 3 == 0:
    return False
```
- **Explanation**: If `n` is divisible by 2 or 3 (and greater than 3), it cannot be a prime number.
- **Example**:
  - `n = 4`: `4 % 2 == 0` → **Not prime** (`False`)
  - `n = 9`: `9 % 3 == 0` → **Not prime** (`False`)

#### 4. **Check for Factors Using 6k ± 1 Optimization**
```python
i = 5
while i * i <= n:
    if n % i == 0 or n % (i + 2) == 0:
        return False
    i += 6
return True
```
- **Explanation**:
  - **Why 6k ± 1?**: All prime numbers greater than 3 can be expressed in the form of `6k ± 1`, where `k` is a positive integer. This is because any integer can be represented as one of the following forms: `6k`, `6k ± 1`, `6k ± 2`, or `6k + 3`. Numbers of the form `6k`, `6k + 2`, and `6k + 4` are divisible by 2, while `6k + 3` is divisible by 3.
  - **Loop Logic**:
    - Start with `i = 5` (which is `6*1 - 1`).
    - Check if `n` is divisible by `i` or `i + 2` (i.e., `5` or `7` in the first iteration).
    - Increment `i` by 6 each time to check the next potential pair of factors (`11` and `13`, `17` and `19`, etc.).
    - Continue this process until `i * i` exceeds `n`. If no divisors are found, `n` is prime.

- **Example 1**: Checking if `n = 29` is prime.
  1. **Initial Checks**:
     - `29 > 1` → Continue
     - `29 > 3` → Continue
     - `29 % 2 != 0` and `29 % 3 != 0` → Continue
  2. **6k ± 1 Loop**:
     - `i = 5`
       - `5 * 5 = 25 <= 29`
       - `29 % 5 != 0` and `29 % 7 != 0` → Continue
     - `i = 11`
       - `11 * 11 = 121 > 29` → Exit loop
  3. **Result**: No divisors found → **Prime** (`True`)

- **Example 2**: Checking if `n = 25` is prime.
  1. **Initial Checks**:
     - `25 > 1` → Continue
     - `25 > 3` → Continue
     - `25 % 2 != 0` and `25 % 3 != 0` → Continue
  2. **6k ± 1 Loop**:
     - `i = 5`
       - `5 * 5 = 25 <= 25`
       - `25 % 5 == 0` → **Not prime** (`False`)
  3. **Result**: Divisor found (`5`) → **Not Prime** (`False`)

- **Example 3**: Checking if `n = 37` is prime.
  1. **Initial Checks**:
     - `37 > 1` → Continue
     - `37 > 3` → Continue
     - `37 % 2 != 0` and `37 % 3 != 0` → Continue
  2. **6k ± 1 Loop**:
     - `i = 5`
       - `5 * 5 = 25 <= 37`
       - `37 % 5 != 0` and `37 % 7 != 0` → Continue
     - `i = 11`
       - `11 * 11 = 121 > 37` → Exit loop
  3. **Result**: No divisors found → **Prime** (`True`)

#### **Why This Approach is Efficient**

1. **Reduced Number of Checks**:
   - The original `is_prime(n)` function checks all numbers from `2` to `n-1`, resulting in **O(n)** time complexity per check.
   - The refactored function reduces this by:
     - Eliminating even numbers and multiples of 3 upfront.
     - Only checking factors up to the square root of `n` (`i * i <= n`), which reduces the number of iterations significantly.
     - Using the `6k ± 1` pattern to skip unnecessary checks, further optimizing the process.

2. **Time Complexity**:
   - **Original Approach**: **O(n²)** overall for finding all primes up to `limit` (since each prime check is **O(n)** and you perform it **n** times).
   - **Refactored Approach**: **O(n * sqrt(n))** overall for finding all primes up to `limit`, which is a substantial improvement.

### **Visual Flowchart**

To further clarify, here's a simplified flowchart of how `is_prime_refactored(n)` works:

1. **Start**
2. **Is `n <= 1`?**
   - **Yes** → Not prime (`False`)
   - **No** → Continue
3. **Is `n <= 3`?**
   - **Yes** → Prime (`True`)
   - **No** → Continue
4. **Is `n` divisible by `2` or `3`?**
   - **Yes** → Not prime (`False`)
   - **No** → Continue
5. **Initialize `i = 5`**
6. **While `i * i <= n`**:
   - **Is `n` divisible by `i` or `i + 2`?**
     - **Yes** → Not prime (`False`)
     - **No** → Increment `i` by `6` and repeat
7. **End Loop** → Prime (`True`)

### **Comparison with Original `is_prime(n)` Function**

#### **Original Function**
```python
def is_prime(n):
    if n <= 1:
        return False
    for j in range(2, n):
        if n % j == 0:
            return False
    return True
```

- **Process**:
  - Checks divisibility from `2` up to `n-1`.
  - **Inefficient** because it performs many unnecessary checks, especially for large `n`.

- **Example**: Checking `n = 29`
  - Iterates through `j = 2` to `28`, performing 27 checks.
  - No divisors found → **Prime** (`True`)

#### **Refactored Function**
- **Process**:
  - Eliminates even numbers and multiples of 3 first.
  - Checks divisibility only up to `sqrt(n)`, and only for numbers of the form `6k ± 1`.
  - **Efficient** as it significantly reduces the number of checks.

- **Example**: Checking `n = 29`
  - Initial checks pass (`n > 3`, not divisible by `2` or `3`).
  - Checks `i = 5` and `i + 2 = 7` → Neither divides `29`.
  - Increments `i` to `11`; since `11 * 11 > 29`, stops.
  - **Prime** (`True`) with only 2 additional checks.


### **Summary with Examples**

Let's summarize the function's behavior with a few more examples to solidify understanding.

### **Example 4: `n = 15`**
1. **Initial Checks**:
   - `15 > 1` → Continue
   - `15 > 3` → Continue
   - `15 % 2 != 0` but `15 % 3 == 0` → **Not Prime** (`False`)

### **Example 5: `n = 17`**
1. **Initial Checks**:
   - `17 > 1` → Continue
   - `17 > 3` → Continue
   - `17 % 2 != 0` and `17 % 3 != 0` → Continue
2. **6k ± 1 Loop**:
   - `i = 5`
     - `5 * 5 = 25 > 17` → Exit loop
3. **Result**: No divisors found → **Prime** (`True`)

### **Example 6: `n = 25`**
1. **Initial Checks**:
   - `25 > 1` → Continue
   - `25 > 3` → Continue
   - `25 % 2 != 0` and `25 % 3 != 0` → Continue
2. **6k ± 1 Loop**:
   - `i = 5`
     - `5 * 5 = 25 <= 25`
     - `25 % 5 == 0` → **Not Prime** (`False`)
3. **Result**: Divisor found → **Not Prime** (`False`)

The `is_prime_refactored(n)` function enhances efficiency through:

1. **Early Elimination**: Quickly rules out non-prime numbers by checking divisibility by 2 and 3.
2. **Optimized Iteration**: Only checks potential factors up to the square root of `n` and leverages the `6k ± 1` pattern to minimize unnecessary checks.
3. **Reduced Complexity**: Decreases the number of iterations from **O(n)** to approximately **O(√n)** per prime check, leading to significant performance gains, especially for large numbers.

By implementing these optimizations, the refactored function ensures faster execution while maintaining accurate prime detection.

2. **`find_primes_refactored(limit)`**:
   - This function generates a list of all prime numbers up to the specified `limit`.
   - **Logic**:
     - Similar to the original function, it initializes an empty list called `primes`.
     - It iterates through every number from `0` to `limit - 1`, checking if each number is prime using the `is_prime_refactored` function. If a number is found to be prime, it is appended to the `primes` list.

   **Example**:
   ```python
   def find_primes_refactored(limit):
       primes = []
       for number in range(limit):
           if is_prime_refactored(number):
               primes.append(number)
       return primes
   ```

3. **Execution Timing**:
   - Similar to the original code, it measures the execution time of the `find_primes_refactored` function by capturing the start time before execution and the end time afterward. The difference provides the total time taken to find the prime numbers.

   **Example**:
   ```python
   start_time = time.time()
   primes_refactored = find_primes_refactored(10000)
   end_time = time.time()
   ```

4. **Output**:
   - Finally, the code prints the total number of prime numbers found and the execution time taken to perform the computation.

   **Example**:
   ```python
   print(f"Refactored Total Prime Numbers: {len(primes_refactored)}")
   print(f"Execution Time (Refactored): {end_time - start_time:.6f} seconds")
   ```

### **Improvements Made**:

1. **Optimized Prime Checking**:
   - The `is_prime_refactored` function utilizes an optimized approach to check for primality, significantly reducing the number of iterations needed for larger numbers compared to the original implementation. The use of the `6k ± 1` rule further improves efficiency.

   **Performance Impact**: This results in a time complexity of O(√n) for checking each prime, leading to a substantial overall performance improvement when finding primes up to `limit`.

2. **Clarity and Readability**:
   - The function and variable names are clear and descriptive, improving code readability. For example, `is_prime_refactored` clearly indicates its purpose of checking for prime numbers.

3. **Simplified Logic**:
   - The logic for checking small numbers and even numbers is straightforward, reducing unnecessary computations and making it easier to understand how the function works.

4. **Maintainability**:
   - The code is structured in a modular fashion, with clear separation of concerns. Each function is focused on a specific task, making it easier to test and modify in the future.

### **Summary of Changes**:

- **Readability**: The refactored code is modular and clear, making it easier to follow and maintain.
- **Performance**: The use of an optimized algorithm for prime checking results in faster execution times, especially for larger input values.
- **Maintainability**: With smaller, well-named functions, the code allows for easier testing and modifications in the future.

In [22]:
import time

def is_prime_refactored(n):
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True

def find_primes_refactored(limit):
    primes = []
    for number in range(limit):
        if is_prime_refactored(number):
            primes.append(number)
    return primes

# Measure execution time for the refactored code
start_time = time.time()
primes_refactored = find_primes_refactored(10000)  # Finding primes up to 10,000
end_time = time.time()

print(f"Refactored Total Prime Numbers: {len(primes_refactored)}")
print(f"Execution Time (Refactored): {end_time - start_time:.6f} seconds")


Refactored Total Prime Numbers: 1229
Execution Time (Refactored): 0.007844 seconds


### Performance Comparison: Original vs. Refactored Code

#### Original Code
- **Execution Time**: Approximately **0.377 seconds**.
- **Structure**: The code directly checks each number for primality using the `prime_check` function within a loop, resulting in multiple calls for each number up to the limit.
- **Complexity**: The `prime_check` function uses an inefficient approach that checks every integer from `2` to `n-1`, resulting in a time complexity of **O(n^2)** overall. This is because for each number checked, it requires `n` iterations to confirm its primality.

#### Refactored Code
- **Execution Time**: **0.007844 seconds**, indicating a significant reduction in execution time.
- **Modular Approach**: The refactored code separates concerns more effectively, using functions like `is_prime_refactored` to encapsulate the logic for checking primality. This enhances readability and makes it easier to maintain.
- **Optimized Logic**: The `is_prime_refactored` function implements a more efficient algorithm, reducing unnecessary checks:
  - It checks for divisibility by `2` and `3` first.
  - It only checks potential factors up to the square root of `n`, which improves efficiency to **O(sqrt(n))** for each check.
  
### Summary of Performance Improvement
- **Execution Time Reduction**: The refactored code demonstrates a dramatic improvement in execution time, achieving nearly instantaneous results.
- **Efficiency of Improved Algorithm**: The use of a more sophisticated primality test drastically reduces the number of iterations required to check if a number is prime.
- **Impact of Function Structure**: By structuring the code into smaller functions, the refactored code not only runs faster but is also easier to read and maintain.

---

### Code Quality

#### Naming Conventions
- **Original Code**:
  - The variable names like `result`, `start_t`, and `end_t` lack clarity and do not convey their purpose effectively.
  - Function names like `prime_check` and `get_primes` are functional but could benefit from more descriptive names that clarify their roles.

- **Refactored Code**:
  - Improved naming conventions such as `is_prime_refactored` and `find_primes_refactored` provide better context about the functionality of each function.
  - Clearer variable names like `primes` help indicate that the variable stores prime numbers.

#### Code Logic
- **Original Code**:
  - The primality check involves unnecessary iterations, making it less efficient. The approach of iterating up to `n-1` is not optimal.
  - Redundant condition checks in `get_primes` increase the overall execution time, as the prime checking logic is executed multiple times for each number.

- **Refactored Code**:
  - The use of checks for small prime numbers (like `2` and `3`) before entering the main loop improves performance.
  - The check iterates only up to the square root of `n`, which minimizes the number of iterations needed and enhances the function’s efficiency.

#### Clarity and Maintainability
- **Original Code**:
  - The compact structure and lack of comments hinder the readability of the code. Users unfamiliar with the logic may struggle to understand its purpose.

- **Refactored Code**:
  - The modular design allows for better organization and clarity. Each function has a single responsibility, making the code easier to follow and modify.
  - The potential for code reuse increases, as smaller, well-defined functions can be called independently.

---

### Conclusion
The refactored code shows substantial improvements over the original implementation in several key areas:

1. **Performance**: The execution time dropped from **0.377 seconds** to **0.007844 seconds** due to the implementation of a more efficient prime-checking algorithm.
2. **Code Quality**: Improved naming conventions and modular structure lead to clearer, more maintainable code.
3. **Overall Efficiency**: The optimization in logic and structure enables the refactored code to handle larger datasets with greater speed and less computational overhead.

By focusing on these aspects, the refactored code not only performs better but also provides a more robust framework for future enhancements and maintenance.

**Note**: While execution times may vary with each run due to factors such as system performance and workload, the refactored code will consistently exhibit lower complexity and improved efficiency compared to the original code. 